sdks/python/apache_beam/io/range_trackers.py

```diff
@@ -115,29 +115,44 @@ def set_current_position(self, record_start):
       self._validate_record_start(record_start, False)
       self._last_record_start = record_start
 
+  def _validate_split_offset(self, split_offset):
+    if split_offset <= self._last_record_start:
+      logging.debug(
+          'Split position %s at %r is not valid: '
+          'already past proposed stop offset', split_offset, self)
+      return False
+    if (split_offset < self.start_position()
+          or split_offset >= self.stop_position()):
+      logging.debug(
+          'Split position %s at %s is not valid: '
+          'proposed split position out of range', split_offset, self)
+      return False
+    return True
+
   def try_split(self, split_offset):
     assert isinstance(split_offset, (int, long))
     with self._lock:
       if self._stop_offset == OffsetRangeTracker.OFFSET_INFINITY:
-        logging.debug('refusing to split %r at %d: stop position unspecified',
+        logging.debug('Refusing to split %r at %d: stop position unspecified',
                       self, split_offset)
         return
+
       if self._last_record_start == -1:
         logging.debug('Refusing to split %r at %d: unstarted', self,
                       split_offset)
         return
 
-      if split_offset <= self._last_record_start:
-        logging.debug(
-            'Refusing to split %r at %d: already past proposed stop offset',
-            self, split_offset)
-        return
-      if (split_offset < self.start_position()
-          or split_offset >= self.stop_position()):
-        logging.debug(
-            'Refusing to split %r at %d: proposed split position out of range',
-            self, split_offset)
-        return
+      if not self._validate_split_offset(split_offset):
+        split_offset = self._last_record_start + 1
```

Comment:
Why are we updating the variable split_offset here ? This looks confusing since we are not doing anything with the update split_offset (other than logging).

Reply:
Ah, I see now that we try to re-split at the updated position. But is "self._last_record_start + 1" what you want ? This split will also fail unless last record was 1 byte, no ? 

Why don't we just rely on the caller to do a better split instead of trying to do two splits in the same try_split() call ?

Reply:
> Ah, I see now that we try to re-split at the updated position.

Just for clarification, we are not trying to split twice. The adjust split only happens when the suggested split is not valid.

> But is "self._last_record_start + 1" what you want ? 

The idea here is, if the suggested split is invalid, the OffsetRangeTracker should split as soon as possible. self._last_record_start + 1 is next record start pos, right?

> This split will also fail unless last record was 1 byte, no ?

Based on how OffsetRangeTracker implemented, I think the adjusted split can only fail when self_.last_record_start has reach to the end of the range. Could you please explain more why you think the last record should be 1 byte?

> Why don't we just rely on the caller to do a better split instead of trying to do two splits in the same try_split() call ?

If the caller refers to SDK, the sdk cannot now what's the next position in the range_tracker,  since it depends on the implementation details of one range tracker. 
If the caller refers to a runner, the runner can only suggest a split. However, a runner can also maintain a split logic likes, asking for split first, if sdk refuses to split, then asking for checkpoint. But this is kind of out scope of this PR.